Centered editor layout

isidorn opened the pull request:
Fixes #50936

Brings back centered editor layout support to VSCode.
This is work in progress, creating PR now to potentially get early feedback.

Still needs to be done:
- [x] double check layout of the grid - seems to overflow to the right
- [x] add separator style (due to overflow right separtor is hidden)
- [x] background color of left and right empty areas should have the color of the editor

joaomoreno commented:
@isidorn I would merged `master` into this, since there was an ugly issue which I just fixed this morning.

isidorn commented:
I have merged master.
I am now also centering the layout by pushing empty views, not reparenting html elements - so it is more elegent.

Still needs some polish from my first comment, once I address that I plan to merge this in if nobody objects

joaomoreno commented on src/vs/workbench/browser/parts/editor/editorPart.ts:
```diff
@@ -967,7 +985,11 @@ export class EditorPart extends Part implements EditorGroupsServiceImpl, IEditor
 		// Layout Grid
 		try {
-			this.gridWidget.layout(this.dimension.width, this.dimension.height);
+			if (this.centeredViewLayout) {
```
Isn't there always a `centeredViewLayout`? This will always be true, it seems.

joaomoreno commented on src/vs/workbench/browser/parts/editor/editorPart.ts:
```diff
@@ -709,15 +711,31 @@ export class EditorPart extends Part implements EditorGroupsServiceImpl, IEditor
 		this.doCreateGridControl();
+		this.centeredViewLayout = new CenteredViewLayout(this.container, {
+			element: this.gridWidget.element,
+			layout: size => this.gridWidget.layout(size, this.dimension ? this.dimension.height : this.gridWidget.maximumHeight),
```
This looks super weird. There must always be a `this.dimension`. It's impossible for this view to be asked for layout before `doLayout` has run.

Also, if you want to remove the weirdness, don't simply use `IView` from splitview. Use `IView` from gridview. This one talks about `width` and `height`. Simultaneously, `CenteredViewLayout.layout` would take in both `width` and `height`.

joaomoreno commented on src/vs/base/browser/ui/centered/centeredViewLayout.ts:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	activate(active: boolean): void {
+		this.active = active;
```
You probably also need `if (this.active === active) { return; }` before this.

joaomoreno commented on src/vs/base/browser/ui/centered/centeredViewLayout.ts:
```diff
@@ -0,0 +1,58 @@
+			};
+
+			this.splitView.addView(emptyView, Sizing.Distribute, 0);
+			this.splitView.addView(emptyView, Sizing.Distribute);
```
You need to add two different views... How can that DOM element be put twice in the DOM?

joaomoreno commented on src/vs/base/browser/ui/centered/centeredViewLayout.ts:
```diff
@@ -0,0 +1,58 @@
+		this.active = active;
+		if (this.active) {
+			const emptyView = {
+				element: $('.'),
```
`$('.')` doesn't mean anything. If you wan a div, use `$('div')`. Though you probably should use something descriptive like `$('.centered-layout-margin')`.

joaomoreno commented on src/vs/base/browser/ui/centered/centeredViewLayout.ts:
```diff
@@ -0,0 +1,58 @@
+			this.splitView.addView(emptyView, Sizing.Distribute);
+		} else {
+			this.splitView.removeView(0);
+			this.splitView.removeView(1);
```
Why not disposing the splitview altogether and just put the provided view directly as a child of the container?

joaomoreno commented on src/vs/workbench/browser/parts/editor/editorPart.ts:
```diff
@@ -1009,6 +1031,9 @@ export class EditorPart extends Part implements EditorGroupsServiceImpl, IEditor
 		}
+		if (this.centeredViewLayout) {
+			this.centeredViewLayout = dispose(this.centeredViewLayout);
+		}
```
All you need is `this.centeredViewLayout.dispose()`.

isidorn commented:
@joaomoreno thanks for the review, I have addressed your feedback. There are still some nit-bits to take care of.

What I find a bit ugly is that for some methods the `editorPart` goes directly to the `gridWidget` and for some it goes to the `centeredLayout` instead which forwards to the `gridWidget`.

joaomoreno commented on src/vs/base/browser/ui/centered/centeredViewLayout.ts:
```diff
@@ -3,60 +3,93 @@
+			element: this.view.element,
+			maximumSize: this.view.maximumWidth,
+			minimumSize: this.view.minimumWidth,
+			onDidChange: Event.None,
```
You should expose `this.view.onDidChange` here.

joaomoreno commented on src/vs/base/browser/ui/centered/centeredViewLayout.ts:
```diff
@@ -3,60 +3,93 @@
+		this.view = view;
+		if (this.splitView) {
+			this.splitView.removeView(1);
+			this.splitView.addView(this.getView(), Sizing.Distribute, 1);
```
There is `splitview.distributeViewSizes()`.

joaomoreno commented on src/vs/base/browser/ui/centered/centeredViewLayout.ts:
```diff
@@ -3,60 +3,93 @@
-		if (this.active) {
-			const emptyView = {
-				element: $('.'),
+		if (active) {
```
You should still make a `if (active === !!this.splitview) { return; }`, so you don't get any NPEs.

isidorn commented:
@joaomoreno thanks, I have addressed that feedback in the latest commit.
There are still some issues with the watermark that I want to address before merging in
